examples_tests/28.BSDFValidator/BSDFValidatorApp.cpp

```diff
@@ -50,34 +53,125 @@ BSDFValidatorApp::BSDFValidatorApp(IrrlichtDevice& device)
+    } PACK_STRUCT;
+#include "irr/irrunpack.h"
+
+    // Create mesh.
+    VertexStruct vertices[4] =
+    {
+        { -0.5f, -0.5f, 0.f },
+        { 0.5f, -0.5f, 0.f },
+        { 0.5f, 0.5f, 0.f },
+        { -0.5f, 0.5f, 0.f }
+    };
+
+    uint16_t indices[6] =
+    {
+        0, 1, 2,
+        2, 3, 0
+    };
+
+    // Upload mesh to the GPU
+    auto upStreamBuff = m_VideoDriver->getDefaultUpStreamingBuffer();
+    const void* dataToPlace[2] = { vertices, indices };
+    uint32_t offsets[2] = { video::StreamingTransientDataBufferMT<>::invalid_address,video::StreamingTransientDataBufferMT<>::invalid_address };
+    uint32_t alignments[2] = { sizeof(decltype(vertices[0u])),sizeof(decltype(indices[0u])) };
+    uint32_t sizes[2] = { sizeof(vertices),sizeof(indices) };
+    upStreamBuff->multi_place(2u, (const void* const*)dataToPlace, offsets, sizes, alignments);
+
+    if (upStreamBuff->needsManualFlushOrInvalidate())
+    {
+        auto upStreamMem = upStreamBuff->getBuffer()->getBoundMemory();
+        m_VideoDriver->flushMappedMemoryRanges({ video::IDriverMemoryAllocation::MappedMemoryRange(upStreamMem, offsets[0],sizes[0]),video::IDriverMemoryAllocation::MappedMemoryRange(upStreamMem,offsets[1],sizes[1]) });
+    }
+
+    m_Mesh = core::make_smart_refctd_ptr<video::IGPUMeshBuffer>();
+
+    {
+        // Define vertex attribute layout. 
+        auto desc = m_VideoDriver->createGPUMeshDataFormatDesc();
+
+        {
+            auto buff = core::smart_refctd_ptr<video::IGPUBuffer>(upStreamBuff->getBuffer());
+
+            // Position attribute.
+            desc->setVertexAttrBuffer(core::smart_refctd_ptr(buff), asset::EVAI_ATTR0, asset::EF_R32G32B32_SFLOAT, sizeof(VertexStruct), 0);
+
+            desc->setIndexBuffer(std::move(buff));
+        }
+
+        m_Mesh->setIndexBufferOffset(offsets[1]);
+        m_Mesh->setIndexType(asset::EIT_16BIT);
+        m_Mesh->setIndexCount(6);
+        m_Mesh->setMeshDataAndFormat(std::move(desc));
+    }
+
+    m_VideoDriver->setTransform(video::E4X3TS_WORLD, core::matrix4x3());
+
+    upStreamBuff->multi_free(2u, (uint32_t*)&offsets, (uint32_t*)&sizes, m_VideoDriver->placeFence());
+
+}
+
+BSDFValidatorApp::~BSDFValidatorApp()
+{
+    delete m_ShaderManager;
 }
 
 void BSDFValidatorApp::RenderGUI()
 {
     m_GUI->render();
 }
 
+void BSDFValidatorApp::RenderMesh()
+{
+    m_Material.MaterialType = m_ShaderManager->GetShader();
+    m_VideoDriver->setMaterial(m_Material);
+
+    /* !!!....TEMPORARY....!!! */
+    m_VideoDriver->drawMeshBuffer(m_Mesh.get());
+}
+
 void BSDFValidatorApp::EventFunctionDefinitionBrowse(const CEGUI::EventArgs& e)
 {
     const auto p = m_GUI->openFileDialog(s_FunctionDefinitionFileDialogTitle, m_FunctionDefinitionFileDialogFilters);
     if (p.first)
     {
-        LoadDefinitions(p.second);
+        std::string functionDefinitions = LoadDefinitions(p.second);
+        if (!functionDefinitions.empty())
```

Comment:
you might want to check for shader compile fail.

Reply:
Isn't it already done deep within `COpenGLDriver::addHighLevelShaderMaterial` somewhere in `source/Irrlicht/COpenGLSLMaterialRenderer.cpp` ?

I was getting compile errors when I appended illegal function definitions to the shader. Is it worth doing it again?

Reply:
if `COpenGLDriver::addHighLevelShaderMaterial` returns <0 then the shader failed to build.